### PR TITLE
Add topic name to the topic status

### PR DESCRIPTION
**Pull request by sknot-rh**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Refactoring

### Description

The K8S KafkaTopic resource does contain `topicName` field. It should not be changed since K8S resource should be bounded to exactly one Kafka Topic. Changing `spec.topicName` brings inconsistency between K8S and Kafka topics.

Putting (and not changing) the `topicName` to the status makes it unambiguous with what `spec.topicName` was topic created.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaTopicStatus.java [hunk 1]
```diff
@@ -19,9 +20,20 @@
         builderPackage = Constants.FABRIC8_KUBERNETES_API
 )
 @JsonInclude(JsonInclude.Include.NON_NULL)
-@JsonPropertyOrder({ "conditions", "observedGeneration" })
+@JsonPropertyOrder({ "topicName", "conditions", "observedGeneration" })
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaTopicStatus.java:23]: I think it should be last after conditions and observed geenration?

**Review by scholzj — COMMENTED** [review]
Can you please explain a bit more in the PR description what does it mean and why is it important? Is this related to the operator renaming topics on its own? Or what does it relate to?

**Comment by scholzj** [comment]
So what does it do when the `.spec` name is changed? Set some error condition and send an event? Or does it revert the value and updates the CR?

**Comment by sknot-rh** [comment]
It does create an event and sets status similar to this 
```
spec:
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
  partitions: 1
  replicas: 1
  topicName: my-topic-renamed
status:
  conditions:
  - lastTransitionTime: 2021-04-09T08:53:23.915275Z
    message: Kafka topics cannot be renamed, but KafkaTopic's spec.topicName has changed.
    reason: IllegalArgumentException
    status: "True"
    type: NotReady
  observedGeneration: 2
  topicName: my-topic
```

**Review by scholzj — APPROVED** [review]

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java [hunk 2]
```diff
@@ -1010,6 +1010,15 @@ protected void observedTopicFuture(KafkaTopic observedTopic) {
                     KafkaTopicStatus kts = new KafkaTopicStatus();
                     StatusUtils.setStatusConditionAndObservedGeneration(topic, kts, result);
 
+
+
```
Comment (tombentley) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java:1014]: No need for 3 blank lines here

**Review by tombentley — COMMENTED** [review]
You've only done the check when we see the Kafka topic change in kube. I think we need similar logic in `getKafkaAndReconcile` (which is on the path when reconciliation is timer-triggered). Best to factor out a method.

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicDiff.java [hunk 3]
```diff
@@ -225,7 +225,7 @@ public static TopicDiff diff(Topic source, Topic target) {
         Objects.requireNonNull(source.getTopicName());
         Objects.requireNonNull(target.getTopicName());
         if (!source.getTopicName().equals(target.getTopicName())) {
-            throw new IllegalArgumentException();
+            throw new IllegalArgumentException("Kafka topics cannot be renamed, but KafkaTopic's spec.topicName has changed.");
```
Comment (tombentley) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicDiff.java:228]: I wonder if we really still need this. I would have thought that we were not detecting name changes before we get as far as doing the diff.

**Review by tombentley — APPROVED** [review]

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicDiff.java (hunk 3 above)
Comment (sknot-rh) [reply topic-operator/src/main/java/io/strimzi/operator/topic/TopicDiff.java:228]: IIRC this path is active in periodical reconciliation.

**Pull request merged**